File: naddu/naddu_2/c/tt.c
#include <string.h>
#include "tt.h"

static uint64_t tt_hash[TT_SIZE];
static uint32_t tt_move[TT_SIZE];
static uint8_t tt_type[TT_SIZE];
static int16_t tt_score[TT_SIZE];
static uint8_t tt_depth[TT_SIZE];

void tt_init(void) {
  tt_clear();
}

void tt_clear(void) {
  memset(tt_type, 0, TT_SIZE);
}

void tt_put(const Pos *pos, int type, int depth, int score, uint32_t move) {
  int index = pos->hash & TT_MASK;

  tt_hash[index] = pos->hash;
  tt_move[index] = move;
  tt_type[index] = type;
  tt_score[index] = score;
  tt_depth[index] = depth;
}

int tt_get(const Pos *pos) {
  uint64_t hash = pos->hash;
  int index = hash & TT_MASK;

  if (tt_hash[index] == hash && tt_type[index]) {
    return index;
  }

  return -1;
}

uint32_t tt_get_move(int index) {
  return tt_move[index];
}

int tt_get_type(int index) {
  return tt_type[index];
}

int tt_get_score(int index) {
  return tt_score[index];
}

int tt_get_depth(int index) {
  return tt_depth[index];
}

int tt_score_to_tt(int score, int ply) {
  if (score > TT_MATE_BOUND)
    return score + ply;
  if (score < -TT_MATE_BOUND)
    return score - ply;
  return score;
}

int tt_score_from_tt(int score, int ply) {
  if (score > TT_MATE_BOUND)
    return score - ply;
  if (score < -TT_MATE_BOUND)
    return score + ply;
  return score;
}

File: naddu/naddu_2/src/tt.js
const TT_EXACT = 1;
const TT_ALPHA = 2;
const TT_BETA = 3;
const TT_MATE_BOUND = 9900;
const TT_SIZE = 1 << 20;
const TT_MASK = TT_SIZE - 1;

const ttHashLo = new Uint32Array(TT_SIZE);
const ttHashHi = new Uint32Array(TT_SIZE);
const ttMove = new Uint32Array(TT_SIZE);
const ttType = new Uint8Array(TT_SIZE);
const ttScore = new Int16Array(TT_SIZE);
const ttDepth = new Uint8Array(TT_SIZE);

const ttStats = {hits: 0};

function ttInitOnce() {
}

function ttClear() {
  ttType.fill(0);
}

function ttPut(pos, type, depth, score, move) {

  const index = pos.hashLo & TT_MASK;

  ttHashLo[index] = pos.hashLo;
  ttHashHi[index] = pos.hashHi;
  ttMove[index] = move;
  ttType[index] = type;
  ttScore[index] = score;
  ttDepth[index] = depth;

}

function ttGet(pos) {

  const hashLo = pos.hashLo;
  const hashHi = pos.hashHi;
  const index = hashLo & TT_MASK;

  if (ttHashLo[index] === hashLo && ttHashHi[index] === hashHi && ttType[index]) {
    return index;
  }

  return -1;
}

function ttGetMove(index) {
  return ttMove[index];
}

function ttGetType(index) {
  return ttType[index];
}

function ttGetScore(index) {
  return ttScore[index];
}

function ttGetDepth(index) {
  return ttDepth[index];
}

function ttScoreToTT(score, ply) {
  if (score > TT_MATE_BOUND)
    return score + ply;
  if (score < -TT_MATE_BOUND)
    return score - ply;
  return score;
}

function ttScoreFromTT(score, ply) {
  if (score > TT_MATE_BOUND)
    return score - ply;
  if (score < -TT_MATE_BOUND)
    return score + ply;
  return score;
}
